File: app/src/main/java/com/skybox/seven/covid/viewmodels/ContactsViewModel.java
package com.skybox.seven.covid.viewmodels;

import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.skybox.seven.covid.model.ContactModel;
import com.skybox.seven.covid.model.ContactRequestModel;
import com.skybox.seven.covid.network.HealthService;
import com.skybox.seven.covid.network.responses.AccessToken;
import com.skybox.seven.covid.network.responses.GenericResponse;
import com.skybox.seven.covid.repository.SharedPreferenceRepository;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ContactsViewModel extends ViewModel {
    private SharedPreferenceRepository preferenceRepository;
    public MutableLiveData<Boolean> contactsRefresh = new MutableLiveData<>();
    private HealthService service;
    public MutableLiveData<List<ContactModel.ContactUsersContacts>> actualContacts = new MutableLiveData<>(new ArrayList<>());
    public MutableLiveData<List<ContactRequestModel.PendingContacts>> pendingContacts = new MutableLiveData<>(new ArrayList<>());
    public MutableLiveData<Boolean> contactsLoading = new MutableLiveData<>();
    public MutableLiveData<Boolean> networkLoading = new MutableLiveData<>(false);

    @ViewModelInject
    public ContactsViewModel(SharedPreferenceRepository sharedPreferenceRepository, HealthService service) {
        this.preferenceRepository = sharedPreferenceRepository;
        this.service = service;
    }

    public String getToken() {
        AccessToken token = preferenceRepository.getToken();
        if (token.getToken() != null) {
            return token.getType() + " " + token.getToken();
        }
        return "";
    }

    public void getAllContacts() {
        contactsLoading.setValue(true);
        networkLoading.setValue(false);
        Call<ArrayList<ContactModel.ContactUsersContacts>> call = service.getAllContacts(getToken());

        call.enqueue(new Callback<ArrayList<ContactModel.ContactUsersContacts>>() {
            @Override
            public void onResponse(Call<ArrayList<ContactModel.ContactUsersContacts>> call, Response<ArrayList<ContactModel.ContactUsersContacts>> response)
            {
                actualContacts.setValue(response.body());
                contactsLoading.setValue(false);
                networkLoading.setValue(false);
            }
            @Override
            public void onFailure(Call<ArrayList<ContactModel.ContactUsersContacts>> call, Throwable t) {
                contactsLoading.setValue(false);
                networkLoading.setValue(true);
            }
        });
    }

    public void generatePendingContactList() {
        contactsLoading.setValue(true);
        networkLoading.setValue(false);
        Call<ArrayList<ContactRequestModel.PendingContacts>> call = service.getPendingContacts(getToken());

        call.enqueue(new Callback<ArrayList<ContactRequestModel.PendingContacts>>() {
            @Override
            public void onResponse(Call<ArrayList<ContactRequestModel.PendingContacts>> call, Response<ArrayList<ContactRequestModel.PendingContacts>> response) {
                pendingContacts.setValue(response.body());
                contactsLoading.setValue(true);
                networkLoading.setValue(false);
            }

            @Override
            public void onFailure(Call<ArrayList<ContactRequestModel.PendingContacts>> call, Throwable t) {
                contactsLoading.setValue(false);
                networkLoading.setValue(true);
            }
        });
    }

    public void acceptContact(int id, int position) {
        networkLoading.setValue(false);
        Call<GenericResponse> call = service.verifyContact(getToken(), id, "accepted");
        call.enqueue(new Callback<GenericResponse>() {
            @Override
            public void onResponse(Call<GenericResponse> call, Response<GenericResponse> response) {
                pendingContacts.getValue().remove(position);
                pendingContacts.setValue(pendingContacts.getValue());
                networkLoading.setValue(true);
            }

            @Override
            public void onFailure(Call<GenericResponse> call, Throwable t) {
                networkLoading.setValue(true);
            }
        });
    }

    public void rejectContact(int id, int position) {
        networkLoading.setValue(false);
        Call<GenericResponse> call = service.verifyContact(getToken(), id, "rejected");
        call.enqueue(new Callback<GenericResponse>() {
            @Override
            public void onResponse(Call<GenericResponse> call, Response<GenericResponse> response) {
                pendingContacts.getValue().remove(position);
                pendingContacts.setValue(pendingContacts.getValue());
                networkLoading.setValue(true);
            }

            @Override
            public void onFailure(Call<GenericResponse> call, Throwable t) {
                networkLoading.setValue(true);
            }
        });
    }
}

File: app/src/main/java/com/skybox/seven/covid/ui/profile/contacts/ContactsViewModel.kt
package com.skybox.seven.covid.ui.profile.contacts

import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.skybox.seven.covid.model.ContactModel.ContactUsersContacts
import com.skybox.seven.covid.model.ContactRequestModel.PendingContacts
import com.skybox.seven.covid.network.HealthService
import com.skybox.seven.covid.network.responses.GenericResponse
import com.skybox.seven.covid.repository.SharedPreferenceRepository
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.schedulers.Schedulers
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class ContactsViewModel @ViewModelInject constructor(private val preferenceRepository: SharedPreferenceRepository,
                                                     private val service: HealthService,
                                                     private val compositeDisposable: CompositeDisposable) : ViewModel() {

    var contactsRefresh = MutableLiveData<Boolean>()
    var actualContacts = MutableLiveData<MutableList<ContactUsersContacts>>(ArrayList())
    var pendingContacts = MutableLiveData<MutableList<PendingContacts>>(ArrayList())
    var contactsLoading = MutableLiveData<Boolean>()
    var networkLoading = MutableLiveData(false)

    val token: String
        get() {
            val token = preferenceRepository.token
            return if (token.token != null) {
                token.type + " " + token.token
            } else ""
        }

    val allContacts: Unit
        get() {
            contactsLoading.value = true
            networkLoading.value = false
            compositeDisposable.add(
                    service.getAllContacts(token)
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe({
                                actualContacts.value = it
                                contactsLoading.value = false
                                networkLoading.setValue(false)
                            }, {
                                contactsLoading.value = false
                                networkLoading.setValue(true)
                            })
            )
        }

    fun generatePendingContactList() {
        contactsLoading.value = true
        networkLoading.value = false
        compositeDisposable.add(
                service.getPendingContacts(token)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({
                            pendingContacts.value = it
                            contactsLoading.value = true
                            networkLoading.value = false
                        }, {
                            contactsLoading.value = false
                            networkLoading.value = true
                        })
        )
    }

    fun acceptContact(id: Int, position: Int) {
        networkLoading.value = false
        compositeDisposable.add(
                service.verifyContact(token, id, "accepted")
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({
                            pendingContacts.value?.removeAt(position)
                            pendingContacts.value = pendingContacts.value
                            networkLoading.setValue(true)
                        }, {
                            networkLoading.setValue(true)
                        })
        )
    }

    fun rejectContact(id: Int, position: Int) {
        networkLoading.value = false
        compositeDisposable.add(
                service.verifyContact(token, id, "rejected")
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({
                            pendingContacts.value?.removeAt(position)
                            pendingContacts.value = pendingContacts.value
                            networkLoading.setValue(true)
                        }, {
                            networkLoading.setValue(true)
                        })
        )
    }

    override fun onCleared() {
        super.onCleared()
        compositeDisposable.dispose()
    }
}